### PR TITLE
Fix indexing in _mapreducedim for OffsetArrays

### PR DESCRIPTION
**Pull request by jishnub**
The destination array was being indexed incorrectly if it had offset indices. This led to the following on nightly:
```julia
julia> using OffsetArrays

julia> r = 5:100;

julia> a = OffsetVector(r, 2);

julia> sum(a, dims=1)
1-element OffsetArray(::Vector{Int64}, 3:3) with eltype Int64 with indices 3:3:
 0

julia> sum(a)
5040
```
The indexing was marked `@inbounds`, so this was not throwing an error. This PR also follows #55329 and only marks the indexing operations as `@inbounds`, omitting the function calls.

### CODE REVIEW

**Comment by jishnub** [comment]
Bump

File: base/reducedim.jl [hunk 1]
```diff
@@ -258,8 +258,9 @@ function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted)
         # use mapreduce_impl, which is probably better tuned to achieve higher performance
         nslices = div(length(A), lsiz)
         ibase = first(LinearIndices(A))-1
-        for i = 1:nslices
-            @inbounds R[i] = op(R[i], mapreduce_impl(f, op, A, ibase+1, ibase+lsiz))
+        for i = range(firstindex(R), length=nslices)
```
Comment (mbauman) [inline base/reducedim.jl:261]: I think this would be even better:

```suggestion
        for i in eachindex(R)
```

Comment (jishnub) [reply base/reducedim.jl:261]: That's certainly better. I hadn't looked into whether the range spans the entire axis

**Review by mbauman — APPROVED** [review]

File: base/reducedim.jl (hunk 1 above)
Comment (mbauman) [reply base/reducedim.jl]: Yeah, I wouldn't have known this had I not been neck deep in reductions lately. That's the biggest advantage here IMV — using eachindex brings so much more clarity to what is happening here, which really isn't that complicated.

**Pull request merged**